@xen-orchestra/xapi/src/vm.js

```diff
@@ -425,18 +425,32 @@ module.exports = class Vm {
     if (srRef !== undefined) {
       query.sr_id = srRef
     }
+
+    let cancel
     if (onVmCreation != null) {
+      const ct = CancelToken.source()
+      cancel = ct.cancel
+
       ignoreErrors.call(
-        this._waitObject(
-          obj => obj != null && obj.current_operations != null && taskRef in obj.current_operations
-        ).then(onVmCreation)
+        this._waitObject(obj => obj != null && obj.current_operations != null && taskRef in obj.current_operations, {
+          cancelToken: ct.token,
+        }).then(vm => {
```

Comment:
The `then` introduce a race condition, the call to `onVmCreation` might be scheduled but not started yet.